dlt/destinations/snowflake/snowflake.py

```diff
@@ -48,44 +53,82 @@
     "VARIANT": "complex"
 }
 
+@with_config(spec=AwsCredentials, sections=(known_sections.DESTINATION, "filesystem", "credentials"))
+def _s3_config(config: AwsCredentials = config.value) -> AwsCredentials:
+    return config
+
+
 class SnowflakeLoadJob(LoadJob, FollowupJob):
     def __init__(
             self, file_path: str, table_name: str, write_disposition: TWriteDisposition, load_id: str, client: SnowflakeSqlClient,
-            stage_name: Optional[str] = None, keep_staged_files: bool = True
+            stage_name: Optional[str] = None, storage_integration: Optional[str] = None, keep_staged_files: bool = True
     ) -> None:
         file_name = FileStorage.get_file_name_from_file_path(file_path)
         super().__init__(file_name)
 
-        with client.with_staging_dataset(write_disposition == "merge"):
+        with client.with_staging_dataset(write_disposition in ["merge", "replace"]):
             qualified_table_name = client.make_qualified_table_name(table_name)
 
-            if stage_name:
+            # extract and prepare some vars
+            bucket_path = CopyFileLoadJob.get_bucket_path(file_path) if CopyFileLoadJob.is_reference_job(file_path) else ""
+            file_name = FileStorage.get_file_name_from_file_path(bucket_path) if bucket_path else file_name
+            from_clause = ""
+            credentials_clause = ""
+            pattern_clause = ""
+            stage_file_path = ""
+
+            # create storage integration stage for reference jobs if defined, for now gcs only
+            if bucket_path and storage_integration:
+                stage_name = client.make_qualified_table_name(stage_name) if stage_name else qualified_table_name
+                bucket = CopyFileLoadJob.get_bucket(file_path)
+                client.execute_sql(f"""
+                    CREATE STAGE IF NOT EXISTS {stage_name}
+                    URL = '{bucket}'
+                    STORAGE_INTEGRATION = {storage_integration};""")
+                from_clause = f"FROM @{stage_name}"
+                # select all files that match the file from our job
+                pattern_clause = f"PATTERN = '.*{file_name}'"
+            # s3 credentials case
+            elif bucket_path:
+                if bucket_path.startswith("s3://"):
+                    s3_config = _s3_config()
+                    credentials_clause = f"""CREDENTIALS=(AWS_KEY_ID='{s3_config.aws_access_key_id}' AWS_SECRET_KEY='{s3_config.aws_secret_access_key}')"""
```

Comment:
```
IAM user: IAM credentials are required. Temporary (aka “scoped”) credentials are generated by AWS Security Token Service (STS) and consist of three components:

AWS_KEY_ID

AWS_SECRET_KEY

AWS_TOKEN

All three are required to access a private/protected bucket. After a designated period of time, temporary credentials expire and can no longer be used. You must then generate a new set of valid temporary credentials.
```
you are not passing AWS_TOKEN. probably you need to get it from boto3. ask GPT-4 :)

Reply:
maybe add it to `AwsCredentials` there's already a `aws_session_token` which we can populate from boto3

Reply:
when does this happen? my tests pass here..